File: sdks/python/docflow/storage/azure.py
"""Azure Blob Storage backend."""

from typing import BinaryIO, List, Optional

from .base import Storage


class AzureStorage(Storage):
    """Azure Blob Storage implementation.
    
    Requires azure-storage-blob: pip install azure-storage-blob
    
    Example:
        >>> storage = AzureStorage(
        ...     account_name="myaccount",
        ...     container_name="documents"
        ... )
        >>> storage.save("doc.pdf", pdf_bytes)
    """

    def __init__(
        self,
        account_name: str,
        container_name: str,
        account_key: Optional[str] = None,
        connection_string: Optional[str] = None,
        prefix: str = "",
    ) -> None:
        """Create Azure Blob storage.
        
        Args:
            account_name: Azure storage account name.
            container_name: Container name.
            account_key: Optional account key (uses DefaultAzureCredential if not provided).
            connection_string: Optional connection string.
            prefix: Optional prefix for all blob names.
        """
        try:
            from azure.storage.blob import BlobServiceClient, ContainerClient
        except ImportError:
            raise ImportError(
                "azure-storage-blob is required: pip install azure-storage-blob"
            )

        self.container_name = container_name
        self.prefix = prefix.rstrip("/")
        self.account_name = account_name

        if connection_string:
            self.blob_service = BlobServiceClient.from_connection_string(connection_string)
        elif account_key:
            account_url = f"https://{account_name}.blob.core.windows.net"
            self.blob_service = BlobServiceClient(account_url, credential=account_key)
        else:
            # Use DefaultAzureCredential
            from azure.identity import DefaultAzureCredential
            account_url = f"https://{account_name}.blob.core.windows.net"
            self.blob_service = BlobServiceClient(account_url, credential=DefaultAzureCredential())

        self.container_client = self.blob_service.get_container_client(container_name)

    def save(self, path: str, data: bytes) -> None:
        """Save data to Azure Blob."""
        blob_name = self._full_key(path)
        blob_client = self.container_client.get_blob_client(blob_name)
        blob_client.upload_blob(data, overwrite=True)

    def save_file(self, path: str, file: BinaryIO) -> None:
        """Save data from a file-like object to Azure Blob."""
        blob_name = self._full_key(path)
        blob_client = self.container_client.get_blob_client(blob_name)
        blob_client.upload_blob(file, overwrite=True)

    def load(self, path: str) -> bytes:
        """Load data from Azure Blob."""
        blob_name = self._full_key(path)
        blob_client = self.container_client.get_blob_client(blob_name)
        try:
            downloader = blob_client.download_blob()
            return downloader.readall()
        except Exception as e:
            raise FileNotFoundError(f"File not found: {path}") from e

    def delete(self, path: str) -> None:
        """Delete file from Azure Blob."""
        blob_name = self._full_key(path)
        blob_client = self.container_client.get_blob_client(blob_name)
        try:
            blob_client.delete_blob()
        except:
            pass  # Ignore if doesn't exist

    def exists(self, path: str) -> bool:
        """Check if file exists in Azure Blob."""
        blob_name = self._full_key(path)
        blob_client = self.container_client.get_blob_client(blob_name)
        return blob_client.exists()

    def list(self, directory: str = "") -> List[str]:
        """List files in Azure Blob prefix."""
        prefix = self._full_key(directory)
        if prefix and not prefix.endswith("/"):
            prefix += "/"

        blobs = self.container_client.list_blobs(name_starts_with=prefix)
        
        files = []
        for blob in blobs:
            rel_path = blob.name[len(prefix):] if prefix else blob.name
            if rel_path and "/" not in rel_path:
                files.append(rel_path)

        return files

    def get_url(self, path: str) -> Optional[str]:
        """Get Azure Blob URL for the file."""
        blob_name = self._full_key(path)
        return f"https://{self.account_name}.blob.core.windows.net/{self.container_name}/{blob_name}"

    def _full_key(self, path: str) -> str:
        """Get full blob name for a relative path."""
        if self.prefix:
            return f"{self.prefix}/{path}"
        return path

File: sdks/go/docflow/storage/azure.go
package storage

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"path"
	"strings"

	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
)

// AzureStorage implements Storage interface for Azure Blob Storage.
type AzureStorage struct {
	client        *azblob.Client
	containerName string
	prefix        string
	accountName   string
}

// AzureConfig contains configuration for Azure Blob storage.
type AzureConfig struct {
	AccountName   string
	AccountKey    string // Optional, uses DefaultAzureCredential if empty
	ContainerName string
	Prefix        string // Optional prefix for all blob names
}

// NewAzureStorage creates a new Azure Blob storage instance.
func NewAzureStorage(cfg AzureConfig) (*AzureStorage, error) {
	if cfg.AccountName == "" {
		return nil, fmt.Errorf("account name is required")
	}
	if cfg.ContainerName == "" {
		return nil, fmt.Errorf("container name is required")
	}

	serviceURL := fmt.Sprintf("https://%s.blob.core.windows.net/", cfg.AccountName)

	var client *azblob.Client
	var err error

	if cfg.AccountKey != "" {
		// Use shared key credential
		cred, err := azblob.NewSharedKeyCredential(cfg.AccountName, cfg.AccountKey)
		if err != nil {
			return nil, fmt.Errorf("failed to create credential: %w", err)
		}
		client, err = azblob.NewClientWithSharedKeyCredential(serviceURL, cred, nil)
		if err != nil {
			return nil, fmt.Errorf("failed to create client: %w", err)
		}
	} else {
		// Use default credential (requires azure-identity)
		client, err = azblob.NewClientWithNoCredential(serviceURL, nil)
		if err != nil {
			return nil, fmt.Errorf("failed to create client: %w", err)
		}
	}

	return &AzureStorage{
		client:        client,
		containerName: cfg.ContainerName,
		prefix:        strings.TrimSuffix(cfg.Prefix, "/"),
		accountName:   cfg.AccountName,
	}, nil
}

// Save stores data at the given path.
func (s *AzureStorage) Save(filePath string, data []byte) error {
	return s.SaveReader(filePath, bytes.NewReader(data))
}

// SaveReader stores data from a reader at the given path.
func (s *AzureStorage) SaveReader(filePath string, reader io.Reader) error {
	blobName := s.fullKey(filePath)

	data, err := io.ReadAll(reader)
	if err != nil {
		return fmt.Errorf("failed to read data: %w", err)
	}

	_, err = s.client.UploadBuffer(context.Background(), s.containerName, blobName, data, nil)
	if err != nil {
		return fmt.Errorf("failed to upload to Azure: %w", err)
	}

	return nil
}

// Load retrieves data from the given path.
func (s *AzureStorage) Load(filePath string) ([]byte, error) {
	blobName := s.fullKey(filePath)

	resp, err := s.client.DownloadStream(context.Background(), s.containerName, blobName, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to download from Azure: %w", err)
	}
	defer resp.Body.Close()

	return io.ReadAll(resp.Body)
}

// LoadReader returns a reader for the data at the given path.
func (s *AzureStorage) LoadReader(filePath string) (io.ReadCloser, error) {
	blobName := s.fullKey(filePath)

	resp, err := s.client.DownloadStream(context.Background(), s.containerName, blobName, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to download from Azure: %w", err)
	}

	return resp.Body, nil
}

// Delete removes the file at the given path.
func (s *AzureStorage) Delete(filePath string) error {
	blobName := s.fullKey(filePath)

	_, err := s.client.DeleteBlob(context.Background(), s.containerName, blobName, nil)
	if err != nil {
		// Ignore if doesn't exist
		return nil
	}

	return nil
}

// Exists checks if a file exists at the given path.
func (s *AzureStorage) Exists(filePath string) (bool, error) {
	blobName := s.fullKey(filePath)

	_, err := s.client.DownloadStream(context.Background(), s.containerName, blobName, nil)
	if err != nil {
		return false, nil
	}

	return true, nil
}

// List returns all files in the given directory.
func (s *AzureStorage) List(dir string) ([]string, error) {
	prefix := s.fullKey(dir)
	if prefix != "" && !strings.HasSuffix(prefix, "/") {
		prefix += "/"
	}

	pager := s.client.NewListBlobsFlatPager(s.containerName, &azblob.ListBlobsFlatOptions{
		Prefix: &prefix,
	})

	var files []string
	for pager.More() {
		resp, err := pager.NextPage(context.Background())
		if err != nil {
			return nil, fmt.Errorf("failed to list blobs: %w", err)
		}

		for _, blob := range resp.Segment.BlobItems {
			relPath := strings.TrimPrefix(*blob.Name, prefix)
			if relPath != "" && !strings.Contains(relPath, "/") {
				files = append(files, relPath)
			}
		}
	}

	return files, nil
}

// GetURL returns the Azure Blob URL for the file.
func (s *AzureStorage) GetURL(filePath string) string {
	blobName := s.fullKey(filePath)
	return fmt.Sprintf("https://%s.blob.core.windows.net/%s/%s", s.accountName, s.containerName, blobName)
}

func (s *AzureStorage) fullKey(filePath string) string {
	if s.prefix == "" {
		return filePath
	}
	return path.Join(s.prefix, filePath)
}
